Update IWCD geometry to match solidworks mechanical design

nickwp opened the pull request.

nickwp committed:
Changes from Purvaja to match mPMT geometry to solidworks design

nickwp committed:
One mPMT per barrel wall face instead of two

pdeperio commented:
Could you include some documentation here? e.g. before/after visualizations, validation plots, effect on key observables, etc. Could be a link to slides too.

nickwp commented:
Here's the most recent sets of summary slides that should be fairly comprehensive:
https://www.hyperk.ca/mpmt/2020-meetings/2020-08-04/summary-so-far-reflector-optimization-i.pdf/view
And an update with some additional plots & numbers:
https://www.hyperk.ca/mpmt/2020-meetings/2020-09-08/optimization-of-reflector-using-wcsim-weekly-update-8th-sept-201920.pdf/view
The optimization is ongoing, so for now the plan is to just use the solidworks mechanical design that is used as the baseline design for the optimization studies.

pdeperio commented on src/WCSimConstructPMT.cc:
```diff
@@ -379,63 +379,25 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
     ////////////////////
     /// 1-PMT support //
     ////////////////////
-    //54mm is position of full PMT support wrt inside of pressure vessel.
```
Where did this all go?

purvaja6 replied:
All the variables are changed according to the CAD drawing for all geometries.

pdeperio commented on src/WCSimConstructPMT.cc:
```diff
@@ -379,63 +379,25 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
-				      tan(mPMT_pmt_openingAngle)*(vessel_radius_curv - mPMT_outer_material_d
-						    -expose - dist_pmt_vessel+id_reflector_z_offset
-						    +id_reflector_height)};
+      G4double ReflectorHolderZ[4] = {0, 22.89, 42.04, 47.18};
```
Should these have CLHEP units?

purvaja6 replied:
Yea I guess. Since the default CLHEP unit is mm I didn't do it explicitly.

pdeperio commented on src/WCSimConstructPMT.cc:
```diff
@@ -379,63 +379,25 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
     // Visualize
     G4VisAttributes* WCPMTVisAtt_sup = new G4VisAttributes(G4Colour(0.3,0.3,0.3));
     WCPMTVisAtt_sup->SetForceSolid(true);
-    logicWCPMTsupport->SetVisAttributes(WCPMTVisAtt_sup);
```
What's this?

purvaja6 replied:
In the default WCSim geometry, there are two separate support structures, one for PMT and one for the reflector. So to adapt the Solidworks design I used polycone shape from GEANT4 with which I could define the entire support structure as one. But this part of the code for all the geometries except WCSim's default geometry is hardcoded keeping the CAD drawings as reference.

pdeperio commented on src/WCSimConstructPMT.cc:
```diff
@@ -379,63 +379,25 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
-						    -expose - dist_pmt_vessel+id_reflector_z_offset
-						    +id_reflector_height)};
+      G4double ReflectorHolderZ[4] = {0, 22.89, 42.04, 47.18};
+      G4double ReflectorHolderR[4] = {39.735, 43.15, 45.89, 46.615};
```
Should these be assigned to variables instead? i.e they no longer correspond to `id_reflector_z_offset`, `id_reflector_height`, `vessel_radius_curv`, etc.?

purvaja6 replied:
So, the above two lines actually represent the polycone shape which I used to define the support structure for the entire PMT where the z values represent the plane, r represents the inner radius and R represents the outer radius. These do not correspond to the variables you have mentioned above. They are different. `id_reflector_z_offset` gives the offset values with respect to the PMT placement. `id_reflector_height` is defined in WCSimDetectorConfigs.cc which is basically the vertical height of the reflector.`vessel_radius_curv` is the radius of curvature of the dome.

pdeperio commented on src/WCSimConstructPMT.cc:
```diff
@@ -451,7 +413,7 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
 
     new G4PVPlacement(0,
-		      G4ThreeVector(0, 0, position_z_offset),
+		      G4ThreeVector(0, 0, 272.36), // value of the z=0 plane with respect to the origin
```
CLHEP units? Variable name instead?

purvaja6 replied:
So this is the placement of the support. I hardcoded it based on the CAD that I have but I am sure we can assign some variable which would have this z value. And since default CLHEP is mm, I didn't show it explicitly.

pdeperio commented on src/WCSimDetectorConfigs.cc:
```diff
@@ -515,18 +515,18 @@ void WCSimDetectorConstruction::SetNuPrismShort_mPMTGeometry()
-	id_reflector_z_offset = 4.8*CLHEP::mm;       //from KM3Net CAD drawings
-	id_reflector_angle = 48.*CLHEP::deg;         // Need to be remeasured for different PMT curvature 
-	mPMT_pmt_openingAngle = 8.7*CLHEP::deg;     // for hex: 8.5deg
+	id_reflector_height = 9.29*CLHEP::mm;        // for a radius of 7.25mm, for hex: 5.4mm (radius of 6mm)
```
Are all these comments still applicable?

purvaja6 replied:
I don't know about the previous comments because some indicate things that I am not familiar with. But the comments with 'me:' are my comments just to indicate things more clearly.

pdeperio commented on src/WCSimDetectorConfigs.cc:
```diff
@@ -515,18 +515,18 @@ void WCSimDetectorConstruction::SetNuPrismShort_mPMTGeometry()
 	vessel_radius_curv = 342.*CLHEP::mm;  //needs to include the vessel thickness, as we construct from outside inwards.
 	vessel_radius = 254.*CLHEP::mm;
-	dist_pmt_vessel = 2*CLHEP::mm;      
+	dist_pmt_vessel = 8*CLHEP::mm;
```
Looks like these changes (and below) are only applied in `SetNuPrismShort_mPMTGeometry`; but how should we keep it synced with the other detectors, e.g. `SetNuPrismBeamTest_mPMTGeometry` and later `HyperK_mPMT`? i.e. a single mPMT class that's shared across detectors?

purvaja6 replied:
I think for that we can change these values similar to `SetNuPrismShort_mPMTGeometry` in all the other detectors. I haven't done much on that but I am sure it will be similar.

nickwp commented:
At the moment, for specific questions about the code you know as much as me (I haven't had time to look through it all) so I'll ask Purvaja if she has time to comment, otherwise I'll try to go through all the changes and work out what's happening.
I think most of these changes are about the reflector holder part of the model. From what I remember, the changes were made so that instead of building the reflector holder as two solid objects that align perfectly according to the PMT and reflector geometry, which doesn't leave any freedom for the changes to the reflector design that she was studying, the holder is now a single solid object with geometry built manually. The old design also used a solid shape that was not compatible with the realistic sold shape for the holder, so that was changed to something more flexible but requires basically defining all the edge positions that are not just defined by the PMT's properties, but placed at specific locations in space relative to the PMT centre. So the holder is not just the perfect truncated cone shape (or two of them) aligned to the PMT, but the more complicated shape that you see in the solidworks design. So I think now the construction of these objects Purvaja implemented is just transferred directly from how solidworks defines it.

nickwp committed:
Add explicit units where missing, update comments for new geometry

nickwp commented:
I've added explicit units and variable names where requested, and updated comments to accurately reflect the new geometry.
The updated mPMT parameters are copied for the beamtest and tall nuprism configs, but not for the other configs using mPMTs since they used completely different parameters anyway (different to both the solidworks and the old IWCD mPMT configuration, maybe they use the hexagonal configuration rather than concentric?)

akutsuR commented on src/WCSimConstructPMT.cc:
```diff
@@ -138,12 +138,12 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructPMT(G4String PMTName, G4Str
       new G4Sphere("WCPMT",
-		   vessel_radius_curv - mPMT_outer_material_d - pmtModuleHeight, //rMin, 54mm is position of support structure
-		   vessel_radius_curv - mPMT_outer_material_d,          //rMax
+		   vessel_radius_curv - mPMT_outer_material_d - pmtModuleHeight, //rMin = 342 - 10 - 54 = 278 mm, 54mm is position of support structure
```
In the above, `pmtModuleHeight` is  declared with 59.62*CLHEP::mm. So the comment, 342 - 10 - 54 = 278mm, seems to be strange.

purvaja6 replied:
Yes, the comment needs to be changed

pdeperio replied:
@nickwp could you quickly fix this?

pdeperio commented on src/WCSimDetectorConfigs.cc:
```diff
@@ -463,18 +463,18 @@ void WCSimDetectorConstruction::SetNuPrismBeamTest_mPMTGeometry()
 	vessel_radius_curv = 342.*CLHEP::mm;  //needs to include the vessel thickness, as we construct from outside inwards.
 	vessel_radius = 254.*CLHEP::mm;
-	dist_pmt_vessel = 2*CLHEP::mm;      
+	dist_pmt_vessel = 8*CLHEP::mm;      
```
@nickwp Thanks for updating. Did you test compilation in this mode? (Later, we should figure out how to create a shared object to minimize hardcode duplication.)

Then should @laurenanthony2 @akutsuR try to run this branch to ensure it's working as expected?

nickwp replied:
I tested compilation but not running and validating WCTE simulations

pdeperio reviewed (changes requested):
After the small fix and testing request below, then I think this is ready to merge, and tag for production, right?

nickwp committed:
Fix comments regarding mPMT geometry calculations

nickwp committed:
Update default 3 inch PMT dark rate to 1 kHz

nickwp commented:
Fixed the geometry calculation comments and also after discussion with @akutsuR and @thomaslindner updated default 3 inch PMT dark rate to 1 kHz (can still be overridden by mac file)

pdeperio reviewed (approved):
When do you hope to start official production? Is there time for small WCTE test production to validate this? (I guess IWCD is validated already?)

nickwp commented:
Not sure the timeline for the neutrino event production (@akutsuR ?) but for ML hope to start asap as soon as we've resolved what to do about the trigger. I should be able to also do some mini production of WCTE particle gun, with some basic checks for validation over the next few days.

akutsuR commented:
> Not sure the timeline for the neutrino event production (@akutsuR ?) but for ML hope to start asap as soon as we've resolved what to do about the trigger. I should be able to also do some mini production of WCTE particle gun, with some basic checks for validation over the next few days.

For neutrino interaction MC, I would like to start a mini production as I have generated some NEUT vectors. I think the highest priority is for new scattering table generation.

nickwp commented:
Seems that this is now ready to commit for the new ML production
